csrc/id_model/id_model.cpp

```diff
@@ -0,0 +1,292 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <id_model/id_model.h>
+#include <id_model/to_string.h>
+
+#include <device_lower/analysis/trivial_broadcast.h>
+#include <device_lower/lower2device.h>
+#include <device_lower/utils.h>
+#include <disjoint_set.h>
+#include <ir/utils.h>
+#include <root_domain_map.h>
+#include <transform_iter.h>
+
+#include <tuple>
+#include <typeinfo>
+#include <utility>
+
+namespace nvfuser {
+
+namespace {
+
+// Map through loop swizzles, as input/output IterDomains are exact, only the
+// order they're traversed differs.
+void mapThroughLoopSwizzles(ValGraph& graph) {
+  std::vector<Swizzle2D*> all_swizzles;
+
+  for (const auto& expr_set :
+       std::as_const(graph).disjointExprSets().disjointSets()) {
+    auto swizzles_in_expr_set = ir_utils::filterByType<Swizzle2D>(
+        expr_set->vector().begin(), expr_set->vector().end());
+    all_swizzles.insert(
+        all_swizzles.end(),
+        swizzles_in_expr_set.begin(),
+        swizzles_in_expr_set.end());
+  }
+
+  for (auto swizzle : all_swizzles) {
+    if (swizzle->swizzleMode() == SwizzleMode::Loop) {
+      graph.mapVals(swizzle->inX(), swizzle->outX());
+      graph.mapVals(swizzle->inY(), swizzle->outY());
+    }
+  }
+}
+
+} // namespace
+
+IdModel::IdModel(
+    const std::vector<Expr*>& exprs,
+    const std::vector<TensorView*>& additional_tvs,
+    bool allow_self_mapping) {
+  build(exprs, additional_tvs);
+}
+
+IdModel::IdModel(const std::vector<Expr*>& exprs, bool allow_self_mapping)
+    : IdModel(exprs, {}, allow_self_mapping) {}
+
+IdModel::IdModel(Fusion* fusion, bool allow_self_mapping) {
+  std::vector<TensorView*> inputs_and_outputs;
+  {
+    auto inp_tvs = ir_utils::filterByType<TensorView>(fusion->inputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), inp_tvs.begin(), inp_tvs.end());
+  }
+  {
+    auto out_tvs = ir_utils::filterByType<TensorView>(fusion->outputs());
+    inputs_and_outputs.insert(
+        inputs_and_outputs.begin(), out_tvs.begin(), out_tvs.end());
```

Comment:
Is there a good reason to add `out_tvs` to the front? (It's obviously slower than to the back).

Reply:
Good question. I don't think so. @csarofeen, is this intentional?

Reply:
Not sure, either way you could just push the outs to the back first then the inputs to get the same order.